CRM/Core/BAO/FinancialTrxn.php

```diff
@@ -696,4 +696,30 @@ public static function createDeferredTrxn($lineItems, $contributionDetails, $upd
     }
   }
 
+  /**
+   * Format Credit Card type.
+   *
+   * @param array $params
+   *
+   */
+  public static function formatCreditCardDetails(&$params) {
+    $creditCardType = CRM_Utils_Array::value('credit_card_type', $params);
+    if ($creditCardType) {
+      if (!is_numeric($creditCardType)) {
+        $creditCardTypes = CRM_Core_PseudoConstant::get('CRM_Financial_DAO_FinancialTrxn',
+          'card_type',
+          array('labelColumn' => 'name')
+        );
+        $params['card_type'] = array_search($creditCardType, $creditCardTypes);
+      }
+      else {
+        $params['card_type'] = $creditCardType;
+      }
+    }
+    $ccNumber = CRM_Utils_Array::value('credit_card_number', $params);
+    if ($ccNumber) {
+      $params['pan_truncation'] = substr($ccNumber, -4);
```

Comment:
I have handled this in BAO layer since contribution can also be created using api by passing credit_card_number as a params. CompleteTransaction also sends 16 digit credit card number. Therefore i truncated here just before its saved in civicrm_financial_trxn table.